File: scripts/capture-har.py
#!/usr/bin/env python3
"""
HAR Fixtures Capture Script — T0.10
Automates capture of 10 user flows using Playwright
"""

import asyncio
import json
from pathlib import Path
from playwright.async_api import async_playwright

BACKEND_URL = "http://localhost:1432"
FRONTEND_URL = "http://localhost:3000"
FIXTURES_DIR = Path(__file__).parent.parent / "fixtures" / "before"

# Test credentials from backend/.env
ADMIN_EMAIL = "admin@example.com"
ADMIN_PASSWORD = "admin123"


async def capture_flow(name: str, description: str, flow_func):
    """Capture a single user flow as HAR"""
    print(f"\n{'='*60}")
    print(f"Capturing: {name}")
    print(f"Description: {description}")
    print(f"{'='*60}")
    
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context(
            record_har_path=str(FIXTURES_DIR / f"{name}.har"),
            record_har_content="embed"
        )
        page = await context.new_page()
        
        try:
            await flow_func(page)
            print(f"✓ {name} captured successfully")
        except Exception as e:
            print(f"✗ {name} failed: {e}")
        finally:
            await context.close()
            await browser.close()


async def flow_01_login(page):
    """Flow 1: Login"""
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    # Check if already logged in
    if "dashboard" in page.url or "login" not in page.url.lower():
        # Logout first
        try:
            await page.click('button:has-text("Logout")', timeout=2000)
            await page.wait_for_url("**/login", timeout=5000)
        except:
            pass
    
    # Fill login form
    await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
    await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
    
    # Submit
    await page.click('button[type="submit"]')
    
    # Wait for redirect
    await page.wait_for_load_state("networkidle", timeout=10000)


async def flow_02_dashboard_load(page):
    """Flow 2: Dashboard Load"""
    # Login first
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    # If not logged in, login
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to dashboard
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    # Wait for dashboard widgets to load
    await asyncio.sleep(2)


async def flow_03_providers_crud(page):
    """Flow 3: Providers CRUD"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to providers page
    try:
        await page.click('a:has-text("Providers"), a[href*="provider"]')
        await page.wait_for_load_state("networkidle")
    except:
        # Try direct URL
        await page.goto(f"{FRONTEND_URL}/providers")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_04_keys_crud(page):
    """Flow 4: Keys CRUD"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to keys page
    try:
        await page.click('a:has-text("Keys"), a:has-text("API Keys"), a[href*="key"]')
        await page.wait_for_load_state("networkidle")
    except:
        await page.goto(f"{FRONTEND_URL}/keys")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_05_settings_save(page):
    """Flow 5: Settings Save"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to settings
    try:
        await page.click('a:has-text("Settings"), a[href*="setting"]')
        await page.wait_for_load_state("networkidle")
    except:
        await page.goto(f"{FRONTEND_URL}/settings")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_06_usage_chart(page):
    """Flow 6: Usage Chart"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to usage/analytics
    try:
        await page.click('a:has-text("Usage"), a:has-text("Analytics"), a[href*="usage"]')
        await page.wait_for_load_state("networkidle")
    except:
        await page.goto(f"{FRONTEND_URL}/usage")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_07_chat_stream(page):
    """Flow 7: Chat Stream"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to chat/test page
    try:
        await page.click('a:has-text("Chat"), a:has-text("Test"), a[href*="chat"]')
        await page.wait_for_load_state("networkidle")
    except:
        await page.goto(f"{FRONTEND_URL}/chat")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_08_oauth_start(page):
    """Flow 8: OAuth Start"""
    # Login
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Navigate to OAuth providers
    try:
        await page.click('a:has-text("OAuth"), a[href*="oauth"]')
        await page.wait_for_load_state("networkidle")
    except:
        await page.goto(f"{FRONTEND_URL}/oauth")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(2)


async def flow_09_oauth_callback(page):
    """Flow 9: OAuth Callback (Placeholder)"""
    # This requires actual OAuth flow completion
    # Capture what we can
    await page.goto(f"{FRONTEND_URL}/oauth/callback")
    await page.wait_for_load_state("networkidle")
    await asyncio.sleep(1)


async def flow_10_logout(page):
    """Flow 10: Logout"""
    # Login first
    await page.goto(FRONTEND_URL)
    await page.wait_for_load_state("networkidle")
    
    if "login" in page.url.lower():
        await page.fill('input[name="email"], input[type="email"]', ADMIN_EMAIL)
        await page.fill('input[name="password"], input[type="password"]', ADMIN_PASSWORD)
        await page.click('button[type="submit"]')
        await page.wait_for_load_state("networkidle")
    
    # Logout
    try:
        await page.click('button:has-text("Logout"), a:has-text("Logout")')
        await page.wait_for_load_state("networkidle")
    except:
        # Try direct API call
        await page.goto(f"{BACKEND_URL}/api/logout")
        await page.wait_for_load_state("networkidle")
    
    await asyncio.sleep(1)


async def main():
    """Capture all flows"""
    print("\n" + "="*60)
    print("HAR Fixtures Capture — T0.10")
    print("="*60)
    print(f"Backend URL: {BACKEND_URL}")
    print(f"Frontend URL: {FRONTEND_URL}")
    print(f"Output directory: {FIXTURES_DIR}")
    print("="*60)
    
    # Ensure fixtures directory exists
    FIXTURES_DIR.mkdir(parents=True, exist_ok=True)
    
    flows = [
        ("01-login", "Email/password authentication", flow_01_login),
        ("02-dashboard-load", "Initial page render with data", flow_02_dashboard_load),
        ("03-providers-crud", "List, create, update, delete providers", flow_03_providers_crud),
        ("04-keys-crud", "List, create, update, delete API keys", flow_04_keys_crud),
        ("05-settings-save", "Update user settings", flow_05_settings_save),
        ("06-usage-chart", "View usage statistics", flow_06_usage_chart),
        ("07-chat-stream", "Initiate streaming chat request", flow_07_chat_stream),
        ("08-oauth-start", "Begin OAuth flow", flow_08_oauth_start),
        ("09-oauth-callback", "Complete OAuth flow", flow_09_oauth_callback),
        ("10-logout", "End session", flow_10_logout),
    ]
    
    for name, description, flow_func in flows:
        await capture_flow(name, description, flow_func)
        await asyncio.sleep(1)  # Brief pause between flows
    
    print("\n" + "="*60)
    print("HAR Capture Complete!")
    print("="*60)
    
    # Verify all files
    print("\nVerifying HAR files...")
    for name, _, _ in flows:
        har_file = FIXTURES_DIR / f"{name}.har"
        if har_file.exists():
            try:
                with open(har_file) as f:
                    json.load(f)
                size = har_file.stat().st_size
                print(f"✓ {name}.har ({size:,} bytes)")
            except json.JSONDecodeError:
                print(f"✗ {name}.har (INVALID JSON)")
        else:
            print(f"✗ {name}.har (NOT FOUND)")
    
    print("\n" + "="*60)
    print("Next steps:")
    print("1. Review captured HARs in fixtures/before/")
    print("2. Redact sensitive data if needed")
    print("3. Commit to repo: git add fixtures/before/*.har")
    print("4. Update TASK_STATUS.md: T0.10 → DONE")
    print("="*60)


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/capture-har.js
/**
 * HAR Fixtures Capture Script - T0.10
 * Captures HTTP Archive files for 10 critical user flows
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const FRONTEND_URL = 'http://localhost:3000';
const BACKEND_URL = 'http://localhost:1432';
const FIXTURES_DIR = path.join(__dirname, '../fixtures/before');

// Test credentials
const TEST_EMAIL = 'admin@example.com';
const TEST_PASSWORD = 'admin123';

async function main() {
  console.log('Starting HAR capture for 10 user flows...');
  console.log('Frontend:', FRONTEND_URL);
  console.log('Backend:', BACKEND_URL);
  console.log('Output:', FIXTURES_DIR);
  console.log('');

  const browser = await chromium.launch({ headless: true });
  
  try {
    // Flow 1: Login
    await captureFlow(browser, '01-login', async (page) => {
      console.log('Capturing 01-login...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      // Check if already logged in, if so logout first
      const loginBtn = await page.$('button:has-text("Login"), a:has-text("Login")');
      const logoutBtn = await page.$('button:has-text("Logout"), a:has-text("Logout")');
      
      if (logoutBtn && !loginBtn) {
        await logoutBtn.click();
        await page.waitForLoadState('networkidle');
      }
      
      // Try to find login form
      const emailInput = await page.$('input[type="email"], input[name="email"], input[placeholder*="email" i]');
      const passwordInput = await page.$('input[type="password"], input[name="password"]');
      
      if (emailInput && passwordInput) {
        await emailInput.fill(TEST_EMAIL);
        await passwordInput.fill(TEST_PASSWORD);
        const submitBtn = await page.$('button[type="submit"], button:has-text("Login"), button:has-text("Sign in")');
        if (submitBtn) {
          await submitBtn.click();
          await page.waitForLoadState('networkidle');
        }
      }
    });

    // Flow 2: Dashboard Load
    await captureFlow(browser, '02-dashboard-load', async (page) => {
      console.log('Capturing 02-dashboard-load...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      await page.waitForTimeout(2000); // Wait for dashboard widgets
    });

    // Flow 3: Providers CRUD
    await captureFlow(browser, '03-providers-crud', async (page) => {
      console.log('Capturing 03-providers-crud...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      // Navigate to providers
      const providersLink = await page.$('a:has-text("Provider"), nav a[href*="provider" i]');
      if (providersLink) {
        await providersLink.click();
        await page.waitForLoadState('networkidle');
      }
      
      await page.waitForTimeout(1000);
    });

    // Flow 4: Keys CRUD
    await captureFlow(browser, '04-keys-crud', async (page) => {
      console.log('Capturing 04-keys-crud...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      const keysLink = await page.$('a:has-text("Key"), nav a[href*="key" i]');
      if (keysLink) {
        await keysLink.click();
        await page.waitForLoadState('networkidle');
      }
      
      await page.waitForTimeout(1000);
    });

    // Flow 5: Settings Save
    await captureFlow(browser, '05-settings-save', async (page) => {
      console.log('Capturing 05-settings-save...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      const settingsLink = await page.$('a:has-text("Setting"), nav a[href*="setting" i]');
      if (settingsLink) {
        await settingsLink.click();
        await page.waitForLoadState('networkidle');
      }
      
      await page.waitForTimeout(1000);
    });

    // Flow 6: Usage Chart
    await captureFlow(browser, '06-usage-chart', async (page) => {
      console.log('Capturing 06-usage-chart...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      const usageLink = await page.$('a:has-text("Usage"), nav a[href*="usage" i]');
      if (usageLink) {
        await usageLink.click();
        await page.waitForLoadState('networkidle');
      }
      
      await page.waitForTimeout(1000);
    });

    // Flow 7: Chat Stream
    await captureFlow(browser, '07-chat-stream', async (page) => {
      console.log('Capturing 07-chat-stream...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      // Look for chat/playground link
      const chatLink = await page.$('a:has-text("Chat"), a:has-text("Playground"), nav a[href*="chat" i]');
      if (chatLink) {
        await chatLink.click();
        await page.waitForLoadState('networkidle');
      }
      
      await page.waitForTimeout(1000);
    });

    // Flow 8: OAuth Start
    await captureFlow(browser, '08-oauth-start', async (page) => {
      console.log('Capturing 08-oauth-start...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      // Look for OAuth buttons
      const oauthBtn = await page.$('button:has-text("Google"), button:has-text("GitHub"), a:has-text("OAuth")');
      if (oauthBtn) {
        // Don't actually click to avoid redirect
        console.log('  Found OAuth button, skipping actual click');
      }
    });

    // Flow 9: OAuth Callback
    await captureFlow(browser, '09-oauth-callback', async (page) => {
      console.log('Capturing 09-oauth-callback...');
      // Simulate OAuth callback URL
      await page.goto(`${FRONTEND_URL}/auth/callback?code=test&state=test`).catch(() => {
        console.log('  OAuth callback page may not exist');
      });
      await page.waitForLoadState('networkidle').catch(() => {});
    });

    // Flow 10: Logout
    await captureFlow(browser, '10-logout', async (page) => {
      console.log('Capturing 10-logout...');
      await page.goto(FRONTEND_URL);
      await page.waitForLoadState('networkidle');
      
      const logoutBtn = await page.$('button:has-text("Logout"), a:has-text("Logout"), button:has-text("Sign out")');
      if (logoutBtn) {
        await logoutBtn.click();
        await page.waitForLoadState('networkidle');
      }
    });

    console.log('\n✅ All HAR files captured successfully!');
    
  } catch (error) {
    console.error('Error during HAR capture:', error);
    throw error;
  } finally {
    await browser.close();
  }
}

async function captureFlow(browser, flowName, flowFn) {
  const context = await browser.newContext({
    recordHar: {
      path: path.join(FIXTURES_DIR, `${flowName}.har`),
      content: 'embed'
    }
  });
  
  const page = await context.newPage();
  
  try {
    await flowFn(page);
  } catch (error) {
    console.error(`  Error in ${flowName}:`, error.message);
  } finally {
    await context.close();
  }
}

main().catch(console.error);
